glommio/src/error.rs

```diff
@@ -17,40 +17,75 @@ pub type Result<T, V> = std::result::Result<T, GlommioError<V>>;
 /// Resource Type used for errors that `WouldBlock` and includes extra
 /// diagnostic data for richer error messages.
 #[derive(Debug)]
+#[non_exhaustive]
 pub enum ResourceType<T> {
     /// Semaphore resource that includes the requested and available shares
-    /// as debugging metadata.
+    /// as debugging metadata for the [`Semaphore`](crate::sync::Semaphore) type.
     Semaphore {
         /// Requested shares
         requested: u64,
         /// Available semaphore shares
         available: u64,
     },
 
-    /// Lock type
+    /// Lock variant for reporting errors from the [`RwLock`](crate::sync::RwLock) type.
     RwLock,
 
-    /// Channel
+    /// Channel variant for reporting errors from [`local_channel`](crate::channels::local_channel) and
+    /// [`shared_channel`](crate::channels::shared_channel) channel types.
     Channel(T),
+
+    /// File variant used for reporting errors for the Buffered ([`BufferedFile`](crate::io::BufferedFile))
+    /// and Direct ([`DmaFile`](crate::io::DmaFile)) file I/O variants.
+    File {
+        /// Specific message for the error
+        msg: String,
+    },
 }
 
 /// Error variant for executor queues.
 #[derive(Debug, Clone, Copy, Eq, PartialEq)]
+#[non_exhaustive]
 pub enum QueueErrorKind {
     /// Queue is still active
     StillActive,
     /// Queue is not found
     NotFound,
 }
 
-#[derive(Error)]
 /// Composite error type to encompass all error types glommio produces.
+///
+/// Single error type that will be produced by any public Glommio API
+/// functions. Contains a generic type that is only used for the [ `Channel` ](ResourceType::Channel)
+/// variants. In other cases it can just be replaced with the unit type `()`
+/// and ignored. The variants are broken up into a few common categories such as
+/// [`Closed`](GlommioError::Closed) and [`WouldBlock`](GlommioError::WouldBlock)
+/// as well as a generic [`IoError`](GlommioError::IoError) and an error dedicated to
+/// executor queue errors ([`QueueError`](GlommioError::QueueError)).
+///
+/// # Examples
+///
+/// ```
+/// use glommio::{GlommioError, ResourceType};
+///
+/// fn will_error() -> Result<(), GlommioError<()>> {
+///     Err(GlommioError::WouldBlock(ResourceType::File {
+///         msg: "Error reading a file".to_string(),
+///     }))?
+/// }
+/// assert!(will_error().is_err());
+///
+/// ```
+#[derive(Error)]
+#[non_exhaustive]
```

Comment:
why are we marking this as non-exhaustive? I think it is healthy to have a catch-all statement in most error-related issues

Reply:
So actually these can go.. but may I make a spirited defense of them if we ever release a 1.0 or this gains massive popularity.. In the ecosystem this is used to save yourself from making what would be considered a breaking change to add an error variant. That all said I am fine with removing it at the stage we are at... though we may want to consider it in the future if we start running into stronger stability requirements from the community. The rust community is maybe a little too obsessed with Sem-ver ;-).. for now going to remove.